pkg/flow/internal/controller/node_custom_component.go

```diff
@@ -0,0 +1,328 @@
+package controller
+
+import (
+	"context"
+	"fmt"
+	"path"
+	"reflect"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/grafana/agent/component"
+	"github.com/grafana/agent/component/module"
+	"github.com/grafana/agent/pkg/flow/config"
+	"github.com/grafana/agent/pkg/flow/logging/level"
+	"github.com/grafana/river/ast"
+	"github.com/grafana/river/vm"
+)
+
+// CustomComponentNode is a controller node which manages a custom component.
+//
+// CustomComponentNode manages the underlying custom component and caches its current
+// arguments and exports.
+type CustomComponentNode struct {
+	id                ComponentID
+	globalID          string
+	label             string
+	componentName     string
+	importLabel       string
+	declareLabel      string
+	nodeID            string // Cached from id.String() to avoid allocating new strings every time NodeID is called.
+	managedOpts       module.ModuleComponentOptions
+	moduleController  ModuleController
+	OnBlockNodeUpdate func(cn BlockNode) // Informs controller that we need to reevaluate
+
+	GetCustomComponentConfig func(*CustomComponentNode) (*CustomComponentConfig, error) // Retrieve the custom component config.
+
+	mut     sync.RWMutex
+	block   *ast.BlockStmt // Current River block to derive args from
+	eval    *vm.Evaluator
+	managed *module.ModuleComponent // Inner managed custom component
+	args    component.Arguments     // Evaluated arguments for the managed component
+
+	// NOTE(rfratto): health and exports have their own mutex because they may be
+	// set asynchronously while mut is still being held (i.e., when calling Evaluate
+	// and the managed custom component immediately creates new exports)
+
+	healthMut  sync.RWMutex
+	evalHealth component.Health // Health of the last evaluate
+	runHealth  component.Health // Health of running the component
+
+	exportsMut sync.RWMutex
+	exports    component.Exports // Evaluated exports for the managed custom component
+}
+
+// ExtractImportAndDeclareLabels extract an importLabel and a declareLabel from a componentName.
+func ExtractImportAndDeclareLabels(componentName string) (string, string) {
+	parts := strings.Split(componentName, ".")
+	if len(parts) == 0 {
+		return "", ""
+	}
+	// If this is a local declare.
+	importLabel := ""
+	declareLabel := parts[0]
+	// If this is an imported custom component.
+	if len(parts) > 1 {
+		importLabel = parts[0]
+		declareLabel = parts[1]
+	}
+	return importLabel, declareLabel
+}
+
+var _ ComponentNode = (*CustomComponentNode)(nil)
+
+// NewCustomComponentNode creates a new CustomComponentNode from an initial ast.BlockStmt.
+// The underlying managed custom component isn't created until Evaluate is called.
+func NewCustomComponentNode(globals ComponentGlobals, b *ast.BlockStmt, GetCustomComponentConfig func(*CustomComponentNode) (*CustomComponentConfig, error)) *CustomComponentNode {
```

Comment:
nit: you might want to make a type for that function outside of this:

```go
type customComponentConfigFunction func(*CustomComponentNode) (*CustomComponentConfig, error) 
```

Reply:
Also, I'm not sure "config" is the right terminology to use here; this is more looking up the template/definition for the declare component. Can we make the name more descriptive?